### PR TITLE
feat(graphical-selector): init nodes stylings and config graph layout format

### PR DESCRIPTION
**Pull request by gtangelo**
Changes made:
- added stylings to node
- created config for graph (default graph is now below)
- added 400 exception for `getPathFrom` endpoint in BE
- added logic to show/hide all nodes in graph
- added override to view inDev features on prod by local storage

**NOTE: Merging this to dev now so that we their is a base code established for later feature development, so the code might be missing some stuff.**

image

### CODE REVIEW

File: frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx [hunk 1]
```diff
@@ -1,72 +1,132 @@
...
+              const edges = graph.getEdges();
+              nodes.forEach((n) => n.show());
+              edges.forEach((e) => e.show());
+            }}
+          >
+            Show all nodes and edges
+          </button>
+        </>
+      )}
+      <S.Wrapper ref={ref} />
```
Comment (Peedee2002) [inline frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx:125]: Wait does this fix the duplicate on reload bug? I noticed that but couldn't be fucked fixing it

File: frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx [hunk 2]
```diff
@@ -1,72 +1,132 @@
...
 import PageTemplate from "components/PageTemplate";
+import S from "./styles";
 
 const GraphicalSelector = () => {
   const { programCode, specs } = useSelector((state) => state.degree);
 
-  const [courses, setCourses] = useState([]);
-  const [courseEdges, setCourseEdges] = useState([]);
+  const [graph, setGraph] = useState(null);
+  const [loading, setLoading] = useState(true);
```
Comment (Peedee2002) [inline frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx:12]: I feel like react had a better way to do this now, but I forget what it was

File: frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx [hunk 3]
```diff
@@ -1,72 +1,132 @@
...
-        modes: {
-          default: ["drag-canvas", "drag-node"],
-        },
-        layout: {
-          type: "force",
-          preventOverlap: true,
+
+  const initialiseGraph = (courses, courseEdges) => {
+    const container = ref.current;
+    const width = container.scrollWidth;
```
Comment (Peedee2002) [inline frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx]: These variables aren't used often enough to warrant making them a variable

File: frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx [hunk 4]
```diff
@@ -1,72 +1,132 @@
...
+    const { structure } = (
+      await axios.get(`/programs/getStructure/${programCode}/${specs.join("+")}`)
+    ).data;
+    const courseList = (
+      Object.values(structure)
+        .flatMap((specialisation) => Object.values(specialisation)
+          .filter((spec) => typeof spec === "object" && spec.courses && !spec.type.includes("rule"))
+          .flatMap((spec) => Object.keys(spec.courses)))
+    );
+    const res = await Promise.all(courseList.map((c) => axios.get(`/courses/getPathFrom/${c}`).catch((e) => e)));
```
Comment (Peedee2002) [inline frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx:82]: Oh this looks like my shitty code oops

File: frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx (hunk 1 above)
Comment (gtangelo) [reply frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx:125]: It doesn't fix the duplicate thing unfortunately. might be just an antv thing rip

File: frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx (hunk 2 above)
Comment (leonardo-fan) [reply frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx:12]: Was it this https://reactjs.org/docs/react-api.html#reactsuspense? Doesn't seem like they support fetching yet :(

**Review by leonardo-fan — APPROVED** [review]
lgtm - just one thing though, when I click a node to hide it, I'm unable to drag the canvas until I press show all again. But given that you just want to merge for a base to work off, up to you what you want to do 😄

**Comment by gtangelo** [comment]
> just one thing though, when I click a node to hide it, I'm unable to drag the canvas until I press show all again. But given that you just want to merge for a base to work off, up to you what you

Hmm interesting will keep that in mind for later on!

**Pull request merged**